keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,504 @@
+
+## Goals
+
+- To guarantee isolation of memory and hugepages to single NUMA node for containers belonging to the Guaranteed Pod.
+
+- To provide topology hint of memory and hugepage resources to the Topology Manager, which will allow Topology Manager to allocate memory, hugepages, CPUs, and PCI devices on the same NUMA node.
+
+## Non-Goals
+
+- Supporting cross-NUMA affinity is out of scope.
+
+- Isolating memory & hugepages to a NUMA node for multiple containers is out of scope.
+
+- Updating K8s scheduler and Pod spec is out of scope at this point.
+
+- This proposal only focuses on Linux based systems.
+
+## User Stories
+
+### Story 1 : Networking Acceleration using DPDK
+
+- The system such as real-time trading system and 5G system, which requires networking acceleration, uses DPDK to guarantee low latency of packet processing. DPDK (Data Plane Development Kit) is set of libraries to accelerate packet processing on userspace. DPDK requires dedicated resources (such as exclusive CPU, hugepages, and DPDK-compatile Network Interface Card) and alignment of resources to sigle NUMA node to prevent unexpected performance degradation due to inter-node communication overhead. For this reason, there should be a way to guarantee a resource reservation of memory and hugepage as well as other computing resources (e.g., CPU core and NIC) from a single NUMA node for DPDK-based containers.
+
+### Story 2 : Database
+
+- Databases (e.g., Oracle, PostgreSQL, and MySQL) require lots of memory and hugepages resource intensively to access a massive volume of data. To reduce memory access latency and improve its performance dramatically, all resources (CPU core, memory, hugepages, and I/O devices) should be aligned to a single NUMA node.
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+Memory Manager is a proposed component of Kubelet which is a new hint provider of Topology Manager especially for memory and hugepage. It provides NUMA affinity to Topology Manager to guarantee NUMA-aware memory consumption for a container which especially belongs to Guaranteed QoS Pod.
+
+When Guaranteed QoS Pod admission is requested from kubelet, Topology Manager asks NUMA affinity of memory and hugepage from Memory Manager. For each container in the Pod, Memory Manager calculates NUMA affinity based on its internal database (i.e. `Memory Map`) which stores a usage of memory and hugepage resource for already-deployed containers. And then, it sends the affinity back to the Topology Manager to help Topology Manager to figure out appropriate NUMA node where adequate resources are available. This calculation is performed for all containers in the Pod and if none of containers is rejected, the Pod is finally admitted to be deployed.
+
+Note:
+- Memory Manager only supports single NUMA affinity, because Linux kernel does not have such a mechanism to reserve memory or set memory limit per NUMA node.
```

Comment:
Does this mean that Memory Manager can only support the `single-numa-node` Topology Manager policy? If so, this could cause confusion to the user - (The user could make a reasonable assumption that Memory Manager behaves like other hint providers under `best-effort` and `restricted` policies).